compiler/src/dotty/tools/dotc/core/Variances.scala

```diff
@@ -98,12 +109,50 @@ object Variances {
       Bivariant
   }
 
-  def varianceString(v: Variance): String =
-    if (v is Covariant) "covariant"
-    else if (v is Contravariant) "contravariant"
-    else "invariant"
+  def setStructuralVariances(lam: HKTypeLambda)(implicit ctx: Context): Unit =
+    assert(!lam.isVariantLambda)
+    for param <- lam.typeParams do param.storedVariance = Bivariant
+    object traverse extends TypeAccumulator[Unit] {
+      def apply(x: Unit, t: Type): Unit = t match
+        case t: TypeParamRef if t.binder eq lam =>
+          lam.typeParams(t.paramNum).storedVariance &= varianceFromInt(variance)
+        case _ =>
+          foldOver(x, t)
+    }
+    // Note: Normally, we'd need to repeat `traverse` until a fixpoint is reached.
+    // But since recursive lambdas can only appear in bounds, and bound never have
+    // structural variances, a single traversal is enough.
+    traverse((), lam.resType)
+
+  /** Does variance `v1` conform to variance `v2`?
+   *  This is the case if the variances are the same or `sym` is nonvariant.
+   */
+   def varianceConforms(v1: Int, v2: Int): Boolean =
+    v1 == v2 || v2 == 0
+
+  /** Does the variance of type parameter `tparam1` conform to the variance of type parameter `tparam2`?
+   */
+   def varianceConforms(tparam1: TypeParamInfo, tparam2: TypeParamInfo)(implicit ctx: Context): Boolean =
+    tparam1.paramVariance.isAllOf(tparam2.paramVariance)
+
+  /** Do the variances of type parameters `tparams1` conform to the variances
+   *  of corresponding type parameters `tparams2`?
+   *  This is only the case of `tparams1` and `tparams2` have the same length.
+   */
+  def variancesConform(tparams1: List[TypeParamInfo], tparams2: List[TypeParamInfo])(implicit ctx: Context): Boolean =
+    val needsDetailedCheck = tparams2 match
+      case (_: Symbol) :: _ => true
+      case LambdaParam(tl: HKTypeLambda, _) :: _ => tl.isVariantLambda
```

Comment:
I'm confused, `isVariantLambda` is only true for lambdas with given variances, but if we're comparing two lambdas with structural variances, shouldn't their variances also be checked for conformance ?

Reply:
I don't think so. For instance
 
    [X] => X => X   <:<    [X] => Any

But the variance of the LHS (invariant) does not conform to the variance of the RHS (bivariant).